tests/pubsub/testgossipsubheartbeat.nim

```diff
@@ -0,0 +1,305 @@
+    var score = 100.0
+    for peer in node0.gossipsub[topic]:
+      if peer in node0.mesh[topic]:
+        # Assign scores in starting Mesh
+        peer.score = score
+        score += 100.0
+      else:
+        # Assign scores higher than median to Peers not in starting Mesh and expect them to be grafted
+        peer.score = 800.0
+        expectedGrafts &= peer
+
+    # Then during heartbeat Peers with lower than median scores are pruned and max 2 Peers are grafted
+    await waitForHeartbeat()
+
+    let actualGrafts = node0.mesh[topic].toSeq().filterIt(it notin startingMesh)
+    check:
+      actualGrafts.len == 2
+      actualGrafts.allIt(it in expectedGrafts)
+
+  asyncTest "Fanout maintenance during heartbeat - expired peers are dropped":
+    let
+      numberOfNodes = 10
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true, fanoutTTL = 30.milliseconds)
+        .toGossipSub()
+
+    startNodesAndDeferStop(nodes)
+    await connectNodesStar(nodes)
+
+    # All nodes but Node0 are subscribed to the topic
+    for node in nodes[1 .. ^1]:
+      node.subscribe(topic, voidTopicHandler)
+    await waitForHeartbeat()
+
+    # When Node0 sends a message to the topic
+    let node0 = nodes[0]
+    tryPublish await node0.publish(topic, newSeq[byte](10000)), 1
+
+    # Then Node0 fanout peers are populated
+    await waitForPeersInTable(node0, topic, 6, PeerTableType.Fanout)
+    check:
+      node0.fanout.hasKey(topic) and node0.fanout[topic].len == 6
+
+    # And after heartbeat (60ms) Node0 fanout peers are dropped (because fanoutTTL=30ms)
+    await waitForHeartbeat()
+    check:
+      not node0.fanout.hasKey(topic)
+
+  asyncTest "Fanout maintenance during heartbeat - fanout peers are replenished":
+    let
+      numberOfNodes = 10
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true).toGossipSub()
+      node0 = nodes[0]
+
+    startNodesAndDeferStop(nodes)
+    await connectNodesStar(nodes)
+
+    # All nodes but Node0 are subscribed  to the topic
+    for node in nodes[1 .. ^1]:
+      node.subscribe(topic, voidTopicHandler)
+    await waitForHeartbeat()
+
+    # When Node0 sends a message to the topic
+    tryPublish await node0.publish(topic, newSeq[byte](10000)), 1
+
+    # Then Node0 fanout peers are populated
+    await waitForPeersInTable(node0, topic, 6, PeerTableType.Fanout)
+    check:
+      node0.fanout.hasKey(topic) and node0.fanout[topic].len == 6
+
+    # When peers of Node0 fanout are disconnected
+    let peersToDisconnect = node0.fanout[topic].toSeq()[1 .. ^1].mapIt(it.peerId)
+    await findAndStopPeers(nodes, peersToDisconnect, topic, voidTopicHandler)
+    await waitForPeersInTable(node0, topic, 1, PeerTableType.Fanout)
+
+    # Then Node0 fanout peers are replenished during heartbeat
+    await waitForHeartbeat()
+    check:
+      node0.fanout[topic].len == 4
+      node0.fanout[topic].toSeq().allIt(it.peerId notin peersToDisconnect)
```

Comment:
`node0.fanout[topic].len == 4` might be unnecessary as condition below will imply this

redundant check is okay, but magic value `4` is not okay -> should be `node0.fanout[topic].toSeq().len - peersToDisconnect.len`

----

more bullet proof check would be:

`node0.fanout[topic]` == `old node0.fanout[topic] removing peersToDisconnect`